elastic/beats
# filebeat/inputs/filestream: add metric for messages truncated

## Proposed commit message

While investigating an SDH, I noticed that although we add the truncated label to log fields, there is no feedback on the amount of messages that are truncated. It often happens that almost all messages are truncated because of a misconfigured input.

This PR adds a input metric that counts the total number of truncated messages.

## Checklist


## How to test this PR locally

Create a log file with multiline logs:

```
cat <<EOF > /tmp/in.log
[beat-logstash-some-name-832-2015.11.28] IndexNotFoundException[no such index]
    at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver\$WildcardExpressionResolver.resolve(IndexNameExpressionResolver.java:566)
    at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver.concreteIndices(IndexNameExpressionResolver.java:133)
    at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver.concreteIndices(IndexNameExpressionResolver.java:77)
    at org.elasticsearch.action.admin.indices.delete.TransportDeleteIndexAction.checkBlock(TransportDeleteIndexAction.java:75)
EOF
```

Configure filebeat to parse the multiline logs with `max_lines`:

```yml
filebeat.inputs:
- type: filestream
  id: foo-bar
  enabled: true
  paths:
    - "/tmp/in.log"
  parsers:
    - multiline:
        type: pattern
        pattern: '^\['
        negate: true
        match: after
        max_lines: 3

output.discard:
  enabled: true
http.enabled: true
```

Check the metrics:

```bash
$ curl -XGET 'localhost:5066/inputs/' | jq
[
  {
    "bytes_processed_total": 596,
    "events_processed_total": 1,
    "files_active": 1,
    "files_closed_total": 0,
    "files_opened_total": 1,
    "id": "foo-bar",
    "input": "filestream",
    "messages_read_total": 1,
    "messages_truncated_total": 1,
    "processing_errors_total": 0,
    "processing_time": {
      "histogram": {
        "count": 1,
        "max": 5000516503,
        "mean": 5000516503,
        "median": 5000516503,
        "min": 5000516503,
        "p75": 5000516503,
        "p95": 5000516503,
        "p99": 5000516503,
        "p999": 5000516503,
        "stddev": 0
      }
    }
  }
]
```

## Related issues

-
